### PR TITLE
v2: partially substitute hyper for reqwest and work around Satellite registry

### LINKED ISSUE #44: v2/client: move to async reqwest
Now that reqwest has an [async](https://docs.rs/reqwest/0.9.2/reqwest/async/index.html) module, we should use that instead of plain hyper. Unfortunately, this means switching from rustls to native-tls, thus getting a FFI dependency to openssl (or similar).

---
Progress
- [x] src/v2/blobs.rs
- [x] src/v2/tags.rs
- [x] src/v2/manifest/mod.rs
- [ ] src/v2/auth.rs #95 
- [ ] src/v2/catalog.rs #95
- [ ] src/v2/mod.rs #95
- [ ] src/v2/config.rs #95 
- [ ] src/lib.rs #95
- [ ] src/errors.rs #95

### LINKED ISSUE #89: v2/client: friction around insecure registries and the `is_v2_supported` method
IMHO the result of the `is_v2_supported()` method isn't correct as it is. 
In case of querying a v2-supported registry via plain http, the 301 status code (MOVED) will cause the method to return `false` to indicate that v2 is not supported here. 
The correct behavior would be look at the new URL and if the scheme changed to https, instead of returning false, throw an error which reports that the requested scheme is not supported.

I'll go ahead and call this a bug, but I'd like to know the original motivation behind this in case I'm missing something.

### PR DESCRIPTION
**Pull request by steveeJ**
Fixes #44 
Partially fixes #89.
Closes #90.

### CODE REVIEW

File: src/v2/auth.rs [hunk 1]
```diff
@@ -89,6 +85,16 @@ impl Client {
...
+        // TODO
+        trace!("login called");
+
+        self.login_hyper(scopes)
+    }
+
+    /// Perform registry authentication and return an authenticated token.
+    ///
+    /// On success, the returned token will be valid for all requested scopes.
+    pub fn login_hyper(&self, scopes: &[&str]) -> FutureTokenAuth {
```
Comment (lucab) [inline src/v2/auth.rs]: This can stay private or `pub(crate)`.

File: src/v2/auth.rs [hunk 2]
```diff
@@ -143,47 +149,65 @@ impl Client {
         Box::new(auth)
     }
 
+    /// Ensure the client is authenticated with the registry.
+    pub fn ensure_auth(&self, token: Option<&str>) -> FutureBool {
```
Comment (lucab) [inline src/v2/auth.rs]: I think this is unused at the moment and can be dropped. If you are planning to use it in library consumer, I think it makes more sense to take ownership of `self` and return a `Future<Item = Self>`.

File: src/v2/tags.rs [hunk 3]
```diff
@@ -20,53 +20,66 @@ struct TagsChunk {
...
+        let client = self.clone();
         let base_url = format!("{}/v2/{}/tags/list", self.base_url, name);
 
-        let fres = futures::stream::unfold(Some(String::new()), move |link| {
+        let fres = futures::stream::unfold(Some(String::new()), move |last| {
+            let client = client.clone();
+
             // Stream ends when response has no `Link` header.
-            let last = match link {
+            let last = match last {
```
Comment (lucab) [inline src/v2/tags.rs]: Can we rename one of those two to avoid shadowing the name of the argument?

File: src/v2/tags.rs [hunk 4]
```diff
@@ -20,53 +20,66 @@ struct TagsChunk {
...
+
+                    client
+                        .build_reqwest(reqwest::async::Client::new().get(url.clone()))
+                        .header(header::ACCEPT, "application/json")
+                        .send()
+                        .map_err(|e| Error::from(format!("{}", e)))
                 })
+                .flatten()
                 .and_then(|resp| {
+                    // ensure the status is OK
```
Comment (lucab) [inline src/v2/tags.rs]: I think `reqwest` has a `error_for_status` which should implement this same logic.

File: src/v2/auth.rs [hunk 5]
```diff
@@ -89,6 +85,16 @@ impl Client {
     ///
     /// On success, the returned token will be valid for all requested scopes.
     pub fn login(&self, scopes: &[&str]) -> FutureTokenAuth {
+        // TODO
```
Comment (lucab) [inline src/v2/auth.rs]: I think you forgot the rest of the TODO message here.

File: src/v2/auth.rs (hunk 1 above)
Comment (steveeJ) [reply src/v2/auth.rs]: I'll remove the indirection for now as I haven't implemented the login in reqwest after all.

**Comment by steveeJ** [comment]
@lucab comments addressed and tests fixed if the status goes green ;-)

**Comment by steveeJ** [comment]
@lucab 
I can't reproduce the test freezes locally. Any idea what's different on Travis?

**Comment by lucab** [comment]
@steveeJ not really, but I'm fine with disabling them for now to get this PR through.

**Pull request merged**